File: Addon/WoW Discord RPC/MessageSerializer.lua
local addonName, addonTable = ...
local Addon = _G[addonName]

local Class = {
	Version = 1,
}
Addon.MessageSerializer = Class

function Class:Serialize()
	local bits = Addon.BitBuffer:New()
	bits:WriteUInt(10, self.Version)

	if self:IncludeCharacter() then
		bits:Write(true)
		Addon.CharacterSerializer:Serialize(bits)
	else
		bits:Write(false)
	end

	if self:IncludeActivity() then
		bits:Write(true)
		Addon.ActivitySerializer:Serialize(bits)
	else
		bits:Write(false)
	end

	return bits
end

function Class:IncludeCharacter()
	return true
end

function Class:IncludeActivity()
	return false
end

File: App/core/src/pl/shockah/wowdiscordrpc/comm/MessageSerializer.java
package pl.shockah.wowdiscordrpc.comm;

import javax.annotation.Nonnull;

import pl.shockah.wowdiscordrpc.bin.BitBuffer;
import pl.shockah.wowdiscordrpc.comm.activity.ActivitySerializer;

public class MessageSerializer {
	public static final int version = 1;

	@Nonnull
	private final CharacterSerializer characterSerializer = new CharacterSerializer();

	@Nonnull
	private final ActivitySerializer activitySerializer = new ActivitySerializer();

	@Nonnull
	public Message deserialize(@Nonnull BitBuffer bits) {
		int readVersion = bits.readUInt(10);
		if (readVersion > version)
			throw new IllegalArgumentException("Unknown version number.");

		Message message = new Message();

		if (bits.read())
			message.character = characterSerializer.deserialize(bits, readVersion);
		if (bits.read())
			message.activity = activitySerializer.deserialize(bits, readVersion);

		return message;
	}
}
